### DISCUSSION TITLE
Alternative to WPF's InputManager.Current.PreProcessInput?

### DISCUSSION THREAD

**Discussion by AngryCarrot789** [AvaloniaUI/Avalonia#15368] [Q&A] [answered]
My app uses the `InputManager.PreProcessInput` event in order to process key events on a global level to activate my shortcut system. But I noticed that the API is all internal in avalonia. Is there a way to achieve the same behaviour?

The basic idea is that I want to be able to intercept key events, pointer events, etc., just before the focused control (for key events) or target control (for mouse click/scroll) is able to handle those events. Then, if my shortcut system activates a shortcut, it will handle/cancel the event and stop the control from ever receiving it, or at least, receives it but as cancelled

**Comment by Gillibald** [comment, marked as answer]
You can always intercept the tunnel phase of the key down events at the top level

**Reply by AngryCarrot789** [reply]
That seems to work nicely, thanks
